### PR TITLE
[MUON-1999] Add video progress tracking to BpkVideoPlayer

### PR DESCRIPTION
**Pull request by Nescafemix**
[MUON-1999]

## Description

Exposes playback progress from `BpkVideoPlayer` so consumers can track video position and fire analytics events (e.g. quartile tracking at 25%, 50%, 75%, 100%).

### Changes

- **New `BpkVideoPlayerProgress` data class** — holds `positionMs`, `durationMs`, and a derived `percentage: Float`.
- **`BpkVideoPlayerController.progressState: State<BpkVideoPlayerProgress?>`** — new observable property, updated every ~200 ms while the video is playing. Emits `null` when duration is unknown. Guarantees a final value at 100% when the video ends or loops (via `onPositionDiscontinuity(DISCONTINUITY_REASON_AUTO_TRANSITION)` for loop-safe detection).
- **Demo** — the "Default Controls" story in `VideoPlayerStory` now shows live progress percentage and position below the player.
- **Tests** — integration tests in `BpkVideoPlayerTest` covering progress polling, pause/resume state retention, end-of-video 100% guarantee, and loop boundary detection; unit tests in `BpkVideoPlayerProgressTest` for `percentage` calculation edge cases.
- **README** — new "Tracking playback progress" section with quartile tracking example.

### How to test

1. Open the demo app → `Video player` → `Default Controls`
2. Observe the progress text updating in real time below the player

https://github.com/user-attachments/assets/6bda1fd4-78a7-4cbf-8604-13558fbfdfb6

---

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] Component `README.md`
+ [x] Tests

[MUON-1999]: https://skyscanner.atlassian.net/browse/MUON-1999?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerController.kt [hunk 1]
```diff
@@ -109,39 +115,72 @@ class BpkVideoPlayerController internal constructor(
...
+                val durationMs = exoPlayer.duration
+                progressState.value = if (durationMs > 0L) {
+                    BpkVideoPlayerProgress(positionMs, durationMs)
+                } else {
+                    null
+                }
+                delay(PROGRESS_POLL_INTERVAL_MS.milliseconds)
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Nescafemix) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerController.kt:169]: changed.

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerController.kt [hunk 2]
```diff
@@ -61,6 +63,11 @@ class BpkVideoPlayerController internal constructor(
     private val _isMuted = mutableStateOf(config.startsMuted)
     val isMuted: State<Boolean> get() = _isMuted
 
+    private val _progressState = mutableStateOf<BpkVideoPlayerProgress?>(null)
```
Comment (mldtms) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerController.kt:66]: _progressState is only ever written in startProgressPolling()'s loop and emitFinalProgress(). Neither resetToStart() nor play() (when resuming from Ended) touches it. Maybe -> reset progress explicitly in resetToStart() (and consider doing the same at the start of play() when coming from Ended):

File: app/src/main/java/net/skyscanner/backpack/demo/compose/VideoPlayerStory.kt [hunk 3]
```diff
@@ -63,22 +66,41 @@ fun VideoPlayerDefaultControlsStory(modifier: Modifier = Modifier) {
...
+            )
+        }
+
+        val progressText = controller.progressState.value?.let { p ->
+            "\n${(p.percentage * 100).toInt()}%  •  ${p.positionMs / 1000}s / ${p.durationMs / 1000}s"
+        } ?: "\n"
+
+        BpkText(
+            modifier = Modifier
+                .weight(1f)
```
Comment (mldtms) [inline app/src/main/java/net/skyscanner/backpack/demo/compose/VideoPlayerStory.kt:98]: Why would we use this? Why would we take the whole space that's left? can we just place the text below the video.

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerProgress.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package net.skyscanner.backpack.compose.videoplayer
+
+data class BpkVideoPlayerProgress(
+    val positionMs: Long,
+    val durationMs: Long,
+) {
+    val percentage: Float
+        get() = if (durationMs > 0) positionMs.toFloat() / durationMs.toFloat() else 0f
```
Comment (mldtms) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerProgress.kt]: maybe add `.coerceIn(0f, 1f)` ?

**Review by mldtms — CHANGES_REQUESTED** [review]
_progressState is only ever written in startProgressPolling()'s loop and emitFinalProgress(). Neither resetToStart() nor play() (when resuming from Ended) touches it.

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerController.kt (hunk 2 above)
Comment (Nescafemix) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerController.kt:66]: Fixed. Both resetToStart() and play() (when coming from Ended) now reset _progressState immediately to BpkVideoPlayerProgress(0L, duration) instead of leaving a stale 100% value. If duration is somehow unavailable it falls back to null. Added two integration tests to cover both cases. https://github.com/Skyscanner/backpack-android/pull/2793/commits/ff41ca3069c8808d9c0f45f0cac1b5d3c08bd380

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerProgress.kt (hunk 4 above)
Comment (Nescafemix) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/videoplayer/BpkVideoPlayerProgress.kt]: Done, added .coerceIn(0f, 1f) to percentage. Also added unit tests for both out-of-range cases (position > duration and negative position). https://github.com/Skyscanner/backpack-android/commit/ff41ca3069c8808d9c0f45f0cac1b5d3c08bd380

File: app/src/main/java/net/skyscanner/backpack/demo/compose/VideoPlayerStory.kt (hunk 3 above)
Comment (Nescafemix) [reply app/src/main/java/net/skyscanner/backpack/demo/compose/VideoPlayerStory.kt:98]: The weight(1f) is intentional — it ensures the progress text is always visible on screen regardless of the video height, by splitting the available space between the video and the text within the Column. Without it, on small screens the text could be pushed off screen. Note that this text is only here as a demo to show that progress can be retrieved from the controller — it is not part of the component itself.

**Review by mldtms — APPROVED** [review]
